bitcoin/psbt.c

```diff
@@ -544,6 +545,119 @@ struct amount_sat psbt_output_get_amount(struct wally_psbt *psbt,
 	assert(amount_asset_is_main(&asset));
 	return amount_asset_to_sat(&asset);
 }
+
+static void add(u8 **key, const void *mem, size_t len)
+{
+	size_t oldlen = tal_count(*key);
+	tal_resize(key, oldlen + len);
+	memcpy(*key + oldlen, mem, len);
+}
+
+static void add_type(u8 **key, const u8 num)
+{
+	add(key, &num, 1);
+}
+
+static void add_varint(u8 **key, size_t val)
+{
+	u8 vt[VARINT_MAX_LEN];
+	size_t vtlen;
+	vtlen = varint_put(vt, val);
+	add(key, vt, vtlen);
+}
+
+/**
+ * BIP174:
+ * Type: Proprietary Use Type <tt>PSBT_GLOBAL_PROPRIETARY = 0xFC</tt>
+ * Type: Proprietary Use Type <tt>PSBT_IN_PROPRIETARY = 0xFC</tt>
+ * Type: Proprietary Use Type <tt>PSBT_OUT_PROPRIETARY = 0xFC</tt>
+ */
+#define WALLY_PSBT_PROPRIETARY 0xFC
+#define LIGHTNING_PROPRIETARY_PREFIX "lightning"
+
+u8 *psbt_make_key(const tal_t *ctx, u8 key_subtype, const u8 *key_data)
+{
+	/**
+	 * BIP174:
+	 * Type: Proprietary Use Type <tt>PSBT_GLOBAL_PROPRIETARY = 0xFC</tt>
+	 ** Key: Variable length identifier prefix, followed
+	 *       by a subtype, followed by the key data itself.
+	 *** <tt>{0xFC}|<prefix>|{subtype}|{key data}</tt>
+	 ** Value: Any value data as defined by the proprietary type user.
+	 *** <tt><data></tt>
+	 */
+	u8 *key = tal_arr(ctx, u8, 0);
+	add_type(&key, WALLY_PSBT_PROPRIETARY);
+	add_varint(&key, strlen(LIGHTNING_PROPRIETARY_PREFIX));
+	add(&key, LIGHTNING_PROPRIETARY_PREFIX,
+	    strlen(LIGHTNING_PROPRIETARY_PREFIX));
+	add_type(&key, key_subtype);
+	if (key_data)
+		add(&key, key_data, tal_bytelen(key_data));
+	return key;
+}
+
+bool psbt_input_add_unknown(struct wally_psbt_input *in,
+			    const u8 *key,
+			    const void *value,
+			    size_t value_len)
+{
+	bool result;
+	u8 *data = tal_arr(NULL, u8, value_len);
+	if (!in->unknowns) {
+		if (wally_unknowns_map_init_alloc(1, &in->unknowns) != WALLY_OK)
+			abort();
+	}
+	memcpy(data, memcheck(value, value_len), value_len);
+	result = wally_add_new_unknown(in->unknowns,
+				       cast_const(unsigned char *, key),
+				       tal_bytelen(key),
+				       data, value_len) == WALLY_OK;
+
+	tal_free(data);
+	return result;
```

Comment:
You never document what `result` means; can this actually happen?  If so, document what it means in the header.  If not, don't return bool from this func.